haigha/connection.py

```diff
@@ -405,9 +409,18 @@ def read_frames(self):
         self._channels[0].send_heartbeat()
 
         data = self._transport.read(self._heartbeat)
+        current_time = time.time()
+
         if data is None:
+            # Wait for 2 heartbeat intervals before giving up. See AMQP 4.2.7:
+            # "If a peer detects no incoming traffic (i.e. received octets) for two heartbeat intervals or longer,
+            # it should close the connection"
+            if self._heartbeat and self._last_octet_time and (current_time-self._last_octet_time > 2*self._heartbeat):
```

Comment:
I think since we know this is initialized on connection, we can drop the `and self._last_octet_time`